notebooks/utils/model_upcast_utils.py

```diff
@@ -0,0 +1,193 @@
+import numpy as np
+import openvino as ov
+from openvino._pyopenvino import Node
+from openvino.runtime.op import Parameter, Constant
+from openvino.runtime.opset12 import matmul, convolution
+from openvino.runtime.utils.types import get_element_type
+from typing import List, Dict, Union, Tuple
+
+ops_to_track_map = {
+    'Convolution': convolution,
+    'MatMul': matmul,
+}
+
+
+def get_thresholds_per_op():
+    return {
+        'Convolution': (0.1, 0.003, 0.00),
+        'MatMul': (0.1, 0.04, 0.03),
+    }
+
+
+def partially_upcast_nodes_to_fp32(orig_model: ov.Model, example_input: Union[List, Dict],
+                                   thresholds_per_op: Dict[str, Tuple] = None) -> ov.Model:
+    model = orig_model.clone()  # todo: check if need to clone orig_models
+    nodes_to_track, outs_to_track, _ = insert_results_for_tracked_ops(model)
+    fp16_full_net_infer_values = infer_full_net_in_fp16(nodes_to_track, model, example_input)
+    fp16_infer_values = infer_nodes_in_fp16(nodes_to_track, fp16_full_net_infer_values)
+    fp32_infer_values = infer_nodes_in_fp32(nodes_to_track, fp16_full_net_infer_values)
+    del model
+    new_model = orig_model.clone()
+    mark_nodes_to_upcast_to_fp32(new_model, nodes_to_track, fp16_infer_values, fp32_infer_values, thresholds_per_op)
+    return new_model
+
+
+def insert_results_for_tracked_ops(model) -> (List, List, List):
+    # additional outputs to track inputs and output values of operations of interest
+    nodes_to_track = []
+    outputs = []
+    for i, op in enumerate(model.get_ordered_ops()):
+        if op.get_type_name() not in ops_to_track_map.keys():
+            continue
+        if any(map(lambda input: input.get_node().get_type_name() == 'Result', op.output(0).get_target_inputs())):
+            continue
+        outputs.append(op.output(0))
+        nodes_to_track.append(op)
+
+        node_0 = op.input_value(0).get_node()
+        node_1 = op.input_value(1).get_node()
+        for node in [node_0, node_1]:
+            if node.get_type_name() != 'Constant' and not is_decompression_convert(node):  # for Consts we can take inputs from ov::Model
+                outputs.append(node.output(0))
+
+    orig_outputs = model.outputs.copy()
+    model.add_outputs(outputs)
+    return nodes_to_track, outputs, orig_outputs
```

Comment:
`outputs` and  `orig_outputs` are not used, probably, it makes sense to remove them

Reply:
i suggest to apply it in a separate PR